src/System.Management.Automation/utils/Encoding.cs

```diff
@@ -0,0 +1,399 @@
+
+                case FileEncoding.Default:
+                    result = GetDefaultEncoding();
+                    break;
+
+                case FileEncoding.Oem:
+                    result = ClrFacade.GetOEMEncoding();
+                    break;
+
+                default:
+                    break;
+            }
+
+            return result;
+        }
+
+        /// <summary>
+        /// Retrieve the encoding based on the Cmdlet and the Encoding
+        /// <param name="cmdlet">The cmdlet of interest</param>
+        /// <param name="encoding">The Encoding parameter value</param>
+        /// <returns>System.Text.Encoding</returns>
+        /// </summary>
+        public static Encoding GetEncoding(Cmdlet cmdlet, FileEncoding encoding)
+        {
+            Encoding resolvedEncoding = GetDefaultEncoding();
+            FileEncoding encodingPreference = FileEncoding.Unknown;
+            bool preferenceSetAndValid = false;
+            string name = cmdlet.GetType().FullName.ToLower(CultureInfo.InvariantCulture);
+
+            // An encoding has been specified as a parameter (or the explicit parameter value is "Unknown")
+            if ( encoding != FileEncoding.Unknown )
+            {
+                // If the encoding has been set to WindowsLegacy, we need to look up the actual encoding
+                if ( encoding == FileEncoding.WindowsLegacy )
+                {
+                    resolvedEncoding = GetWindowsLegacyEncoding(name);
+                }
+                else
+                {
+                    resolvedEncoding = GetEncoding(encoding);
+                }
+            }
+            else
+            {
+                // the parameter is not specifically set, so check the preference variable
+                encodingPreference = GetEncodingPreference(cmdlet.Context.SessionState);
+                // If set to unknown, we accept that it is unset
+                preferenceSetAndValid = encodingPreference != FileEncoding.Unknown;
+                // If the encoding preference has been set to WindowsLegacy, we need to look up the actual encoding
+                if ( encodingPreference == FileEncoding.WindowsLegacy )
+                {
+                    resolvedEncoding = GetWindowsLegacyEncoding(name);
+                }
+                else if ( encodingPreference != FileEncoding.Unknown )
+                {
+                    resolvedEncoding = GetEncoding(encodingPreference);
+                }
+                // the final else would be set the encoding to GetDefaultEncoding() which was handled above
+            }
+
+            return resolvedEncoding;
+        }
+
+        /// <summary>
+        /// Given a path to a file, attempt to retrieve the encoding
+        /// <param name="path">The path to a file to inspect for an encoding</param>
+        /// <returns>System.Text.Encoding</returns>
+        /// </summary>
+        public static FileEncoding GetFileEncodingFromFile(string path)
```

Comment:
concrete suggestion please? This code has existed for some time (although not public) and has been used for a long time